### ISSUE TITLE
`User.create` should encode usernames

### ISSUE THREAD

**Issue by jsumners** [NodeBB/NodeBB#5362]
https://github.com/jsumners/nodebb-plugin-sso-discord-alt/issues/5

The `User.create` API method should encode usernames so that names like "Mercurial | corpnewt.com" do not cause an "invalid username" error.

**Comment by julianlam** [comment]
They are encoded for slugs... the issue is that | is not a valid username character :smile:

**Comment by jsumners** [comment]
I know that's the issue. I'm saying it shouldn't be an issue.

**Comment by julianlam** [comment]
@jsumners Right, but I'd argue that running the input through `encodeURIComponent` leads to a degraded user experience.

I can almost guarantee that if we do this, we'll get reports about this being a bug, even if intentional 😄 

We could add `|` to the character whitelist...

**Comment by jsumners** [comment]
My position is that there shouldn't be any concept of an invalid username. Whatever the person decides to name themselves should be valid. Internally, you can encode them however you like, `encodeURIComponent` being one way. But when you present the name to a user, it should be decoded to the way the person behind the name originally typed it.

If you are concerned about how it looks in the URL bar as a "slug", then you can map invalid URI characters to something else specifically for that situation.

**Comment by dhinakg** [comment]
I agree with @jsumners. A lot of plugins crash NodeBB because of invalid characters, and encoding should be added somehow.

**Comment by pichalite** [comment]
@dhinakg which plugins are included in the "lot of plugins"? example?

**Comment by dhinakg** [comment]
For example, nodebb-plugin-sso-discord-alt, and pretty much any other plugin that creates users and can possibly insert an invalid character.

**Comment by nodebb-misty** [comment]
Issue closed due to inactivity.

----

*This is an automated message. If you feel this action was in error, please comment on this issue so it can be looked at again*

**Comment by jsumners** [comment]
That's a fantastic way to deal with issues -- just automatically close them when you don't deal with them.

**Comment by julianlam** [comment]
@jsumners -- what a horrible sentiment to wake up to on a Friday morning.

I understand how frustrated you feel. Misty is a bot and she closes issues that are inactive in a heavy-handed attempt to ensure that our issue tracker is not overwhelmed. The last thing I want to do is get into a situation where we have ten thousand open issues and anyone who posts an issue knows full well it won't even be looked.

I hope in turn you will understand that NodeBB is a labour of love that we maintain for free. Our client-work is separate from our open-source work and it is discouraging that someone would go as far as to publicly shame someone else for not resolving his issue (for free, mind you) fast enough.

You have to understand that these issues are always on the back of my mind. Even when they are closed by Misty. Nice-to-haves and low priority non-bugs that we will get to in time.

I read through your issue. I gave you an honest argument about how it has been working fine so far, and that a workaround would be to not use the pipe character in a username. Your latest reply made me think that I am incorrect in my assessment, and out of embarrassment, I chose not to reply so I could craft a more meaningful response when I had enough time.

I'm finished with this issue. You will have to rely on someone else on this project to overhaul the username validator code. Alternatively, you can make the changes yourself and choose whether or not to propagate these changes back to the core.

**Comment by jsumners** [comment]
@julianlam I don't care how long an issue takes to resolve. But I do care that a bot is in place to simply close issues due to non-activity. That's not a proper way to resolve an issue.

1. It makes it look to the filer that you can just ignore issues and they will "go away".
2. It artificially "improves" your issue count numbers. This makes it look like the project cares more about how well it is perceived to be developed/managed than how it might actually be.
3. It imposes upon you a time limit to properly resolve issues lest the bot come along and close them without resolution.

If you personally don't want to deal with the issue any longer, that's fine. My post in response to the close was merely a direct response to the way open issues are evidently handled. It's not something I have _ever_ seen with any other project. At the very least, a bot comes along an pings those involved at least once to do something; only after a ping, or several, would the bot close the issue.

If an issue is a "nice to have, maybe later" then simply add an "enhancement" or "request-for-assistance" or some such label to it.

**Comment by pitaj** [comment]
Issues get closed due to inactivity because often we reply to an issue and get no further information. Misty does this automatically for precisely this reason, to remind people to continue giving information.

Before I mark this issue with some different label, please give some more examples of different usernames which are currently not allowed and should be by your standard.

**Comment by jsumners** [comment]
@pitaj I don't know of any. But I also don't know what anyone in the world feels like naming themselves on the Internet.

I'm really not sure what other information you need. As @julianlam said, I have given a reasoned description of how I think any username could be handled. Clearly he understood it, he just didn't acknowledge it until today.

Aside from learning your codebase and fix it myself, which I don't really have time to do, I don't know what else to contribute to the issue.

**Comment by nodebb-misty** [comment]
Issue closed due to inactivity.

----

*This is an automated message. If you feel this action was in error, please comment on this issue so it can be looked at again*

**Comment by a632079** [comment]
I find that `（` in `彩阳（新）` will lead this problem too.

**Comment by Dravere** [comment]
I'm really interested in this as well. And I might even have some time to look into this. Do you by chance have some tips/hints for what I should look out. I'm aware of `public/src/utils.js` with `isUserNameValid` and `slugify`. Is there more that should be taken into consideration?

**Comment by Dravere** [comment]
This is getting more complicated the longer I look at it. The slug is mostly used in URLs which I wouldn't see as a problem. Most browsers try to hide the path of the URL to the users anyway. And most laymen don't even now what the URL or the address bar actually is. In those cases I don't think it would be a problem.

But `utils.slugify` is also used in different cases. For example to create titles or mentions. Replacing an @ mentioning with encoded URI components definitely would hurt the user experience. On the other hand one could ask why is the mention even replaced with the slug?

Also I see some problems in backwards compatibility. I think I found places in the code where the slug is regenerated instead of taken from the database. Or the slug is compared in renames to check if a different name was entered.

This is kind of sad. I was really hoping this would be easier to introduce. Especially the backwards compatibility worries me a bit.

@julianlam What are the policies for backward compatibility? Could a `./nodebb upgrade` update all slugs of every user? Or would that be out of scope?

**Comment by psychobunny** [comment]
This issue was brought to my attention recently, and honestly I don't know what the best way to move forward is. I'm closing it so that the watchers can think of reopening or recommenting, and fresh eyes are welcome:

Ideally the issue should contain all of what characters are not allowed and the discussion would be, well, should we allow them? We could have this discourse over on our community as well to canvas additional opinions. @Dravere thanks for thinking about it and sorry for the delay. Happy to assist if you want to take another shot :)

**Comment by oplik0** [comment]
Well, if I had to say what I'd consider the best user experience it would be what @Dravere suggested - slugify used for urls, but the usernames users actually see could be almost all of unicode. That is, everything but control characters (regex: `/^\P{C}+$/iu`), or also without line breaks and paragraph separators (`/^[\P{C}\P{Zl}\P{Zp}]+$/iu`) - this is AFAIK the same set Discord allows (that is, I was unable to find any character from that list Discord couldn't handle. And I actually was trying to because of another issue about what characters should be allowed in usernames: https://github.com/TitanEmbeds/Titan/issues/145)
Mentions would at least show users actual usernames and not slugs, eliminating the problem of weird encodings. This would IMO generally provide better experience even with current allowed character set.

This does theoretically present a problem with mentioning users that utilize weird characters in their usernames (no suggestions and it's hard to type them :V), but quite honestly - if someone has a very weird username they caused themselves to be unmentionable willingly, and NodeBB allows one to disable notifications on mentions anyway, so one could just consider it another way of doing that.

And actually this could improve the mentions experience for people that use non-latin alphabets. AFAIK slugify will delete this kind of characters, so being able to use them might make mentioning people in such locales easier :)

**Comment by jsumners** [comment]
I really don't have any skin in this game any longer, but I'll chime back in to reiterate: display name is different from user name. It's not a huge migration to copy current usernames to a "display name" field and then format the same value for a new "user name" format. @oplik0 mentioned Discord as an example, and it's a perfect one for this. They have `<ascii>#<digits>` as the user name. The value preceding the `#` is used as the default for the display name ("Nickname" in their parlance), but is adjustable on a per "server" basis to anything.

**Comment by oplik0** [comment]
Actually Discord format for usernames is `<unicode>#<digits>` - you are definitely not limited to ascii:
![oplik⌂](https://user-images.githubusercontent.com/25460763/101019911-f4d15900-356d-11eb-8f16-df2243925f7d.png)

But I do think the idea of splitting username (limited to a smaller subset of characters) and display name (all "safe" unicode characters allowed) is great.

This would also make it easier to make plugins that'd, for example, allow for per-category display names.
